File: src/ProxyDB.php
<?php

namespace PhpProxyHunter;

if (!defined('PHP_PROXY_HUNTER')) exit('access denied');

/**
 * Class ProxyDB
 *
 * @package PhpProxyHunter
 */
class ProxyDB
{
  /** @var SQLiteHelper $db */
  private $db;

  /**
   * ProxyDB constructor.
   *
   * @param string|null $dbLocation
   */
  public function __construct(?string $dbLocation = null)
  {
    if (!$dbLocation) $dbLocation = realpath(__DIR__ . '/database.sqlite');
    $this->db = new SQLiteHelper($dbLocation);
  }

  /**
   * Selects a proxy from the database.
   *
   * @param string $proxy
   * @return array
   */
  public function select(string $proxy): array
  {
    return $this->db->select('proxies', '*', 'proxy = ?', [trim($proxy)]);
  }

  /**
   * Get all proxies from the database.
   *
   * @return array
   */
  public function getAllProxies(): array
  {
    return $this->db->select('proxies', '*');
  }

  /**
   * Removes a proxy from the database.
   *
   * @param string $proxy
   */
  public function remove(string $proxy): void
  {
    $this->db->delete('proxies', 'proxy = ?', [trim($proxy)]);
  }

  /**
   * Adds a proxy to the database.
   *
   * @param string $proxy
   */
  public function add(string $proxy): void
  {
    $this->db->insert('proxies', ['proxy' => trim($proxy), 'status' => 'untested']);
  }

  /**
   * Inserts or updates a proxy in the database.
   *
   * @param string $proxy
   * @param string|null $type
   * @param string|null $region
   * @param string|null $city
   * @param string|null $country
   * @param string|null $status
   * @param string|null $latency
   * @param string|null $timezone
   */
  public function update(string $proxy, ?string $type = null, ?string $region = null, ?string $city = null, ?string $country = null, ?string $status = null, ?string $latency = null, ?string $timezone = null): void
  {
    if (empty($this->select($proxy))) {
      $this->add($proxy);
    }
    $data = [];
    if ($city) $data['city'] = $city;
    if ($country) $data['country'] = $country;
    if ($type) $data['type'] = $type;
    if ($region) $data['region'] = $region;
    if ($latency) $data['latency'] = $latency;
    if ($timezone) $data['timezone'] = $timezone;
    if ($status) {
      $data['status'] = $status;
      $data['last_check'] = date(DATE_RFC3339);
    }
    if (!empty($data)) $this->updateData($proxy, $data);
  }

  /**
   * Updates data for a specific proxy.
   *
   * @param string $proxy
   * @param array $data
   */
  public function updateData(string $proxy, array $data = []): void
  {
    if (empty($this->select($proxy))) {
      $this->add($proxy);
    }
    // Remove null and false values
    $data = array_filter($data, function ($value) {
      return $value !== null && $value !== false;
    });
    if (isset($data['status'])) $data['last_check'] = date(DATE_RFC3339);
    if (!empty($data)) $this->db->update('proxies', $data, 'proxy = ?', [trim($proxy)]);
  }

  /**
   * Updates the status of a proxy.
   *
   * @param string $proxy
   * @param string $status
   */
  public function updateStatus(string $proxy, string $status): void
  {
    $this->update(trim($proxy), null, null, null, null, $status, null);
  }

  /**
   * Updates the latency of a proxy.
   *
   * @param string $proxy
   * @param string $latency
   */
  public function updateLatency(string $proxy, string $latency): void
  {
    $this->update(trim($proxy), null, null, null, null, null, $latency);
  }

  /**
   * Gets working proxies from the database.
   *
   * @param int|null $limit The maximum number of working proxies to retrieve. Default is null (no limit).
   * @return array An array containing the working proxies.
   */
  public function getWorkingProxies(?int $limit = null): array
  {
    $whereClause = 'status = ?';
    $params = ['active'];

    // Append the limit clause if the $limit parameter is provided
    $limitClause = ($limit !== null) ? "LIMIT $limit" : '';

    $result = $this->db->select('proxies', '*', $whereClause . ' ' . $limitClause, $params);
    if (!$result) return [];
    return $result;
  }

  /**
   * Gets private proxies from the database.
   *
   * @param int|null $limit The maximum number of private proxies to retrieve. Default is null (no limit).
   * @return array An array containing the private proxies.
   */
  public function getPrivateProxies(?int $limit = null): array
  {
    $whereClause = 'status = ?';
    $params = ['private'];

    // Append the limit clause if the $limit parameter is provided
    $limitClause = ($limit !== null) ? "LIMIT $limit" : '';

    $result = $this->db->select('proxies', '*', $whereClause . ' ' . $limitClause, $params);
    if (!$result) return [];
    return $result;
  }

  /**
   * Gets dead proxies from the database, including those with closed ports.
   *
   * @param int|null $limit The maximum number of dead proxies to retrieve. Default is null (no limit).
   * @return array An array containing the dead proxies.
   */
  public function getDeadProxies(?int $limit = null): array
  {
    $whereClause = 'status = ? OR status = ?';
    $params = ['dead', 'port-closed'];

    // Append the limit clause if the $limit parameter is provided
    $limitClause = ($limit !== null) ? "LIMIT $limit" : '';

    return $this->db->select('proxies', '*', $whereClause . ' ' . $limitClause, $params);
  }

  /**
   * Retrieves untested proxies from the database.
   *
   * @param int|null $limit The maximum number of untested proxies to retrieve. Default is null (no limit).
   * @return array An array containing the untested proxies.
   */
  public function getUntestedProxies(?int $limit = null): array
  {
    $whereClause = 'status IS NULL OR status = "" OR status NOT IN (?, ?, ?)';
    $params = ['active', 'port-closed', 'dead'];

    // Append the limit clause if the $limit parameter is provided
    $limitClause = ($limit !== null) ? "LIMIT $limit" : '';

    return $this->db->select('proxies', '*', $whereClause . ' ' . $limitClause, $params);
  }

  public function countDeadProxies(): int
  {
    $closed = $this->db->count('proxies', 'status = ?', ['port-closed']);
    $dead = $this->db->count('proxies', 'status = ?', ['dead']);
    return $closed + $dead;
  }

  public function countUntestedProxies(): int
  {
    return $this->db->count('proxies', 'status = ? OR status IS NULL OR status = ""', ['']);
  }

  public function countWorkingProxies(): int
  {
    return $this->db->count('proxies', "(status = ?) AND (private = ? OR private IS NULL OR private = '')", [
        'active',
        'false'
    ]);
  }

  public function countPrivateProxies(): int
  {
    return $this->db->count('proxies', "private = ?", ['true']);
  }
}

/**
 * proxy table data class
 */
class Proxy
{
  /** @var int|null */
  public $id = null;

  /** @var string */
  public $proxy;

  /** @var string|null */
  public $latency = null;

  /** @var string|null */
  public $type = null;

  /** @var string|null */
  public $region = null;

  /** @var string|null */
  public $city = null;

  /** @var string|null */
  public $country = null;

  /** @var string|null */
  public $last_check = null;

  /** @var string|null */
  public $anonymity = null;

  /** @var string|null */
  public $status = null;

  /** @var string|null */
  public $timezone = null;

  /** @var string|null */
  public $longitude = null;

  /** @var string|null */
  public $private = null;

  /** @var string|null */
  public $latitude = null;

  /** @var string|null */
  public $lang = null;

  /** @var string|null */
  public $useragent = null;

  /** @var string|null */
  public $webgl_vendor = null;

  /** @var string|null */
  public $webgl_renderer = null;

  /** @var string|null */
  public $browser_vendor = null;

  /** @var string|null */
  public $username = null;

  /** @var string|null */
  public $password = null;

  /**
   * Proxy constructor.
   * @param string $proxy
   * @param string|null $latency
   * @param string|null $type
   * @param string|null $region
   * @param string|null $city
   * @param string|null $country
   * @param string|null $last_check
   * @param string|null $anonymity
   * @param string|null $status
   * @param string|null $timezone
   * @param string|null $longitude
   * @param string|null $private
   * @param string|null $latitude
   * @param string|null $lang
   * @param string|null $useragent
   * @param string|null $webgl_vendor
   * @param string|null $webgl_renderer
   * @param string|null $browser_vendor
   * @param string|null $username
   * @param string|null $password
   * @param int|null $id
   */
  public function __construct(
      string  $proxy,
      ?string $latency = null,
      ?string $type = null,
      ?string $region = null,
      ?string $city = null,
      ?string $country = null,
      ?string $last_check = null,
      ?string $anonymity = null,
      ?string $status = null,
      ?string $timezone = null,
      ?string $longitude = null,
      ?string $private = null,
      ?string $latitude = null,
      ?string $lang = null,
      ?string $useragent = null,
      ?string $webgl_vendor = null,
      ?string $webgl_renderer = null,
      ?string $browser_vendor = null,
      ?string $username = null,
      ?string $password = null,
      ?int    $id = null
  )
  {
    $this->id = $id;
    $this->proxy = $proxy;
    $this->latency = $latency;
    $this->type = $type;
    $this->region = $region;
    $this->city = $city;
    $this->country = $country;
    $this->last_check = $last_check;
    $this->anonymity = $anonymity;
    $this->status = $status;
    $this->timezone = $timezone;
    $this->longitude = $longitude;
    $this->private = $private;
    $this->latitude = $latitude;
    $this->lang = $lang;
    $this->useragent = $useragent;
    $this->webgl_vendor = $webgl_vendor;
    $this->webgl_renderer = $webgl_renderer;
    $this->browser_vendor = $browser_vendor;
    $this->username = $username;
    $this->password = $password;
  }
}

File: android_app/app/src/main/java/com/dimaslanjaka/proxyhunter/data/ProxyDB.kt
package com.dimaslanjaka.proxyhunter.data

import android.content.Context
import timber.log.Timber
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.concurrent.Future

class ProxyDB {
  private val mysql: MySQLHelper?
  private val sqlite: SQLiteHelper?
  private val executor = Executors.newFixedThreadPool(5)

  constructor(
    host: String = "23.94.85.180",
    user: String = "proxyuser",
    pass: String = "proxypassword",
    dbName: String = "myproject",
    port: Int = 3306
  ) {
    this.mysql = MySQLHelper(host, user, pass, dbName, port)
    this.sqlite = null
  }

  constructor(sqlite: SQLiteHelper) {
    this.mysql = null
    this.sqlite = sqlite
  }

  constructor(sqlite_database: String, context: Context) {
    this.mysql = null
    this.sqlite = SQLiteHelper(context,sqlite_database)
  }

  private val isSQLite: Boolean get() = sqlite != null

  private fun logSql(operation: String, sql: String, params: List<Any>) {
    val driver = if (isSQLite) "SQLite" else "MySQL"
    Timber.tag("ProxyDB").d("[%s][%s] %s || %s", driver, operation, sql, params)
  }

  private fun query(operation: String, sql: String, params: List<Any> = emptyList()): Future<List<Map<String, Any?>>> {
    logSql(operation, sql, params)
    return mysql?.query(sql, params) ?: sqlite!!.query(sql, params)
  }

  private fun update(operation: String, sql: String, params: List<Any> = emptyList()): Future<Int> {
    logSql(operation, sql, params)
    return mysql?.update(sql, params) ?: sqlite!!.update(sql, params)
  }

  private fun now(): String = if (isSQLite) "datetime('now')" else "NOW()"
  private fun rand(): String = if (isSQLite) "RANDOM()" else "RAND()"
  private fun insertIgnore(): String = if (isSQLite) "INSERT OR IGNORE" else "INSERT IGNORE"

  fun getAllProxies(limit: Int? = null): Future<List<ProxyItem>> {
    val sql = if (limit != null) "SELECT * FROM proxies LIMIT ?" else "SELECT * FROM proxies"
    val params = if (limit != null) listOf(limit) else emptyList()

    return executor.submit(Callable {
      query("getAllProxies", sql, params).get().map { mapRowToProxyItem(it) }
    })
  }

  fun getWorkingProxies(
    limit: Int = 100,
    offset: Int = 0,
    country: String? = null,
    city: String? = null,
    classification: String? = null,
    type: String? = null
  ): Future<List<ProxyItem>> {
    var sql = "SELECT * FROM proxies WHERE status = 'active'"
    val params = mutableListOf<Any>()

    if (!country.isNullOrBlank()) {
      sql += " AND country = ?"
      params.add(country)
    }
    if (!city.isNullOrBlank()) {
      sql += " AND city = ?"
      params.add(city)
    }
    if (!classification.isNullOrBlank()) {
      sql += " AND classification = ?"
      params.add(classification)
    }
    if (!type.isNullOrBlank()) {
      val normalizedType = type.trim().lowercase()
      if (normalizedType == "ssl") {
        sql += " AND LOWER(COALESCE(https, '')) = 'true'"
      } else {
        if (isSQLite) {
          // Match full token only, so socks4 does not match socks4a (and similar collisions).
          sql += " AND INSTR(',' || REPLACE(REPLACE(LOWER(COALESCE(type, '')), '-', ','), ' ', ',') || ',', ',' || ? || ',') > 0"
          params.add(normalizedType)
        } else {
          sql += " AND LOWER(COALESCE(type, '')) REGEXP ?"
          params.add("(^|[-, ])" + Regex.escape(normalizedType) + "($|[-, ])")
        }
      }
    }

    sql += " ORDER BY last_check DESC LIMIT ? OFFSET ?"
    params.add(limit)
    params.add(offset)

    return executor.submit(Callable {
      query("getWorkingProxies", sql, params).get().map { mapRowToProxyItem(it) }
    })
  }

  fun getDeadProxies(limit: Int = 100, offset: Int = 0): Future<List<ProxyItem>> {
    val sql = "SELECT * FROM proxies WHERE status = 'dead' ORDER BY last_check DESC LIMIT ? OFFSET ?"
    return executor.submit(Callable {
      query("getDeadProxies", sql, listOf(limit, offset)).get().map { mapRowToProxyItem(it) }
    })
  }

  fun getUniqueCountries(): Future<List<String>> {
    val sql =
      "SELECT DISTINCT country FROM proxies WHERE status = 'active' AND country IS NOT NULL AND country != '' ORDER BY country ASC"
    return executor.submit(Callable {
      query("getUniqueCountries", sql).get().mapNotNull { it["country"]?.toString() }
    })
  }

  fun getUniqueCities(country: String? = null): Future<List<String>> {
    var sql = "SELECT DISTINCT city FROM proxies WHERE status = 'active' AND city IS NOT NULL AND city != ''"
    val params = mutableListOf<Any>()
    if (!country.isNullOrBlank()) {
      sql += " AND country = ?"
      params.add(country)
    }
    sql += " ORDER BY city ASC"

    return executor.submit(Callable {
      query("getUniqueCities", sql, params).get().mapNotNull { it["city"]?.toString() }
    })
  }

  fun getUniqueClassifications(): Future<List<String>> {
    val sql =
      "SELECT DISTINCT classification FROM proxies WHERE status = 'active' AND classification IS NOT NULL AND classification != '' ORDER BY classification ASC"
    return executor.submit(Callable {
      query("getUniqueClassifications", sql).get().mapNotNull { it["classification"]?.toString() }
    })
  }

  fun getUniqueTypes(): Future<List<String>> {
    val sql = "SELECT DISTINCT type, https FROM proxies WHERE status = 'active'"
    return executor.submit(Callable {
      val rows = query("getUniqueTypes", sql).get()
      val typeSet = mutableSetOf<String>()
      for (row in rows) {
        val rawType = row["type"] as? String
        val isHttps = row["https"]?.toString()?.lowercase() == "true"

        if (isHttps) typeSet.add("SSL")

        if (rawType != null) {
          rawType.split(",", "-", " ").forEach {
            val trimmed = it.trim().uppercase()
            if (trimmed.isNotEmpty()) typeSet.add(trimmed)
          }
        }
      }

      val preferredOrder = listOf("SSL", "SOCKS4", "SOCKS5", "SOCKS4A", "SOCKS5H")
      val result = preferredOrder.filter { it in typeSet }.toMutableList()
      val others = typeSet.filter { it !in preferredOrder }.sorted()
      result.addAll(others)

      result
    })
  }

  fun getUntestedProxies(limit: Int = 100, randomize: Boolean = true): Future<List<ProxyItem>> {
    var sql = "SELECT * FROM proxies WHERE status IS NULL OR status = 'untested' OR status = ''"

    if (randomize) {
      sql += " ORDER BY ${rand()}"
    }

    sql += " LIMIT ?"
    return executor.submit(Callable {
      query("getUntestedProxies", sql, listOf(limit)).get().map { mapRowToProxyItem(it) }
    })
  }

  fun addProxy(proxy: String): Future<Boolean> {
    val sql = "${insertIgnore()} INTO proxies (proxy, status) VALUES (?, ?)"
    return executor.submit(Callable {
      update("addProxy", sql, listOf(proxy, "untested")).get() > 0
    })
  }

  fun updateStatus(proxy: String, status: String): Future<Int> {
    val sql = "UPDATE proxies SET status = ?, last_check = ${now()} WHERE proxy = ?"
    return update("updateStatus", sql, listOf(status, proxy))
  }

  /**
   * Records a proxy with its detected type and status.
   * If the proxy doesn't exist, it will be inserted.
   */
  fun upsertProxy(proxy: String, type: String, status: String): Future<Boolean> {
    val sql = if (isSQLite) {
      """
            INSERT INTO proxies (proxy, type, status, last_check)
            VALUES (?, ?, ?, datetime('now'))
            ON CONFLICT(proxy) DO UPDATE SET
            type = ?, status = ?, last_check = datetime('now')
        """.trimIndent()
    } else {
      """
            INSERT INTO proxies (proxy, type, status, last_check)
            VALUES (?, ?, ?, NOW())
            ON DUPLICATE KEY UPDATE
            type = ?, status = ?, last_check = NOW()
        """.trimIndent()
    }
    return executor.submit(Callable {
      update("upsertProxy", sql, listOf(proxy, type, status, type, status)).get() > 0
    })
  }

  fun testConnection(): Boolean {
    return try {
      getUniqueCountries().get()
      true
    } catch (e: Exception) {
      false
    }
  }

  private fun mapRowToProxyItem(row: Map<String, Any?>): ProxyItem {
    return ProxyItem(
      id = (row["id"] as? Number)?.toInt() ?: 0,
      proxy = row["proxy"] as? String ?: "",
      latency = row["latency"]?.toString(),
      lastCheck = (row["last_check"] ?: row["lastCheck"])?.toString(),
      type = row["type"]?.toString(),
      region = row["region"]?.toString(),
      city = row["city"]?.toString(),
      country = row["country"]?.toString(),
      timezone = row["timezone"]?.toString(),
      latitude = row["latitude"]?.toString(),
      longitude = row["longitude"]?.toString(),
      anonymity = row["anonymity"]?.toString(),
      https = row["https"]?.toString(),
      status = row["status"]?.toString(),
      private = row["private"]?.toString(),
      lang = row["lang"]?.toString(),
      useragent = row["useragent"]?.toString(),
      webglVendor = (row["webgl_vendor"] ?: row["webglVendor"])?.toString(),
      webglRenderer = (row["webgl_renderer"] ?: row["webglRenderer"])?.toString(),
      browserVendor = (row["browser_vendor"] ?: row["browserVendor"])?.toString(),
      username = row["username"]?.toString(),
      password = row["password"]?.toString(),
      classification = row["classification"]?.toString()
    )
  }

  fun close() {
    mysql?.close()
    sqlite?.close()
    executor.shutdown()
  }
}
